---
title: "IRTG COURSE: INTRODUCTION TO R"
author: "Carl Herrmann & Carlos Ramirez"
site: bookdown::bookdown_site
date: "`r format(Sys.time(), '%d %B, %Y')`"
documentclass: book
output:
  bookdown::gitbook: default
  #bookdown::pdf_book: default
---


```{r, setup, include=FALSE}
knitr::opts_chunk$set(
        warning = FALSE, 
        message = FALSE,
        cache = FALSE,
        fig.align='center'
)
```

# Introduction {#intro}


We will use available data from 10X genomics of Peripheral Blood Mononuclear cells (PBMC). In order
to speed up the process only a subset of 500 cells are going to be processed here. Here, we will
perform the following tasks:

* Definition of a Seurat object from a matrix of counts
* Quality Control of the cell samples
* Filtering
* Dimensional reduction
* Clustering
* Differential expression
* Visualization of markers



## Quantifying gene expression in single cells


Genes are expressed in cells by the synthesis of mRNA molecules. Quantification of gene expression
in single cells consists of counting and mapping sequenced reads to each gene in each cell. The 
outcome of this process is a count matrix, which is a discrete matrix of integers corresponding
to the number of RNA molecules associated to each gene and cell. The count matrix usually has 
genes as rows and cells in columns. 

In the next lines we will load the matrix of UMI counts. Each row of the matrix represent a gene 
and each column represent a cell.


```{r, loading_dataset, eval=FALSE, echo=FALSE, results='hide'}
library(Seurat)
library(rmarkdown)
library(dplyr)

set.seed(333)

#pbmc_url <- 'https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz'
#dir.create('data')
#download.file(pbmc_url, 
#              destfile = 'data/pbmc3k_filtered_gene_bc_matrices.tar.gz')
#untar('data/pbmc3k_filtered_gene_bc_matrices.tar.gz', exdir = 'data/')

# Load the PBMC dataset
#pbmc.data <- Read10X(data.dir = "data/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
#pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc <- subset(pbmc, downsample=500)
#pbmc

#saveRDS(pbmc@assays$RNA@counts,
#        file = 'data/pbmc_10X_500_cells.mtx.rds',
#        compress = TRUE)
pbmc <- pbmc_url  <- 'https://github.com/caramirezal/caramirezal.github.io/raw/master/bookdown-minimal/data/pbmc_10X_500_cells.rds'
pbmc <- readRDS(url(pbmc_url))
dim(pbmc)

saveRDS(pbmc@meta.data,
        file = 'data/pbmc_10X_500_cells.anns.rds',
        compress = TRUE)

## Loading example data
pbmc_url  <- 'https://github.com/caramirezal/caramirezal.github.io/raw/master/bookdown-minimal/data/pbmc_10X_500_cells.mtx.rds'
pbmc.mtx <- readRDS(url(pbmc_url))
dim(pbmc.mtx)
```


As can be seen from the output of the function `dim()` the matrix contains `r nrow(pbmc.mtx)` rows (genes) 
and `r ncol(pbmc.mtx)` columns (cells).




