<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Markers visualization | IRTG COURSE: INTRODUCTION TO R</title>
  <meta name="description" content="9 Markers visualization | IRTG COURSE: INTRODUCTION TO R" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Markers visualization | IRTG COURSE: INTRODUCTION TO R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Markers visualization | IRTG COURSE: INTRODUCTION TO R" />
  
  
  

<meta name="author" content="Carl Herrmann &amp; Carlos Ramirez" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-expression-analysis.html"/>

<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> scRNA-Seq Course Syllabus</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#single-cell-sequencing"><i class="fa fa-check"></i><b>1.1</b> Single Cell Sequencing</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#quantifying-gene-expression-in-single-cells"><i class="fa fa-check"></i><b>1.2</b> Quantifying gene expression in single cells</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html"><i class="fa fa-check"></i><b>2</b> Standard Preprocessing using Seurat</a>
<ul>
<li class="chapter" data-level="2.1" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#creating-seurat-object"><i class="fa fa-check"></i><b>2.1</b> Creating Seurat object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality control</a></li>
<li class="chapter" data-level="4" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>4</b> Feature selection</a></li>
<li class="chapter" data-level="5" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html"><i class="fa fa-check"></i><b>5</b> Dimensionional Reduction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html#normalization"><i class="fa fa-check"></i><b>5.1</b> Normalization</a></li>
<li class="chapter" data-level="5.2" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html#dimensional-reduction"><i class="fa fa-check"></i><b>5.2</b> Dimensional reduction</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="cell-clustering.html"><a href="cell-clustering.html"><i class="fa fa-check"></i><b>6</b> Cell clustering</a></li>
<li class="chapter" data-level="7" data-path="cluster-visualization.html"><a href="cluster-visualization.html"><i class="fa fa-check"></i><b>7</b> Cluster visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cluster-visualization.html"><a href="cluster-visualization.html#umap"><i class="fa fa-check"></i><b>7.1</b> UMAP</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="9" data-path="markers-visualization.html"><a href="markers-visualization.html"><i class="fa fa-check"></i><b>9</b> Markers visualization</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">IRTG COURSE: INTRODUCTION TO R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="markers-visualization" class="section level1" number="9">
<h1><span class="header-section-number">9</span> Markers visualization</h1>
<p>First, we will take top 10 ranked genes based in Log FC and visualize their
expression in clusters using a heatmap representation.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="markers-visualization.html#cb31-1" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> pbmc.degs <span class="sc">%&gt;%</span> </span>
<span id="cb31-2"><a href="markers-visualization.html#cb31-2" aria-hidden="true" tabindex="-1"></a>             <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> </span>
<span id="cb31-3"><a href="markers-visualization.html#cb31-3" aria-hidden="true" tabindex="-1"></a>             <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">wt =</span> avg_log2FC)</span>
<span id="cb31-4"><a href="markers-visualization.html#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(pbmc.seurat, </span>
<span id="cb31-5"><a href="markers-visualization.html#cb31-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">features =</span> top10<span class="sc">$</span>gene) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="course_files/figure-html/heatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>IL-7 is a marker for naive CD4+ T cells, while GZMB is a marker for CD8 T cells.
Then, we can tentatively consider cluster 0 and 2 as CD4 and CD8 T cells,
respectively.</p>
<p>We can visualize additional known canonical markers in order to assign cell
categories.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="markers-visualization.html#cb32-1" aria-hidden="true" tabindex="-1"></a>canonical_markers <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;IL7R&#39;</span>,     <span class="do">## CD4+ cell</span></span>
<span id="cb32-2"><a href="markers-visualization.html#cb32-2" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;CCR7&#39;</span>,     <span class="do">## Naive CD4+ T cell</span></span>
<span id="cb32-3"><a href="markers-visualization.html#cb32-3" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;CD8A&#39;</span>,     <span class="do">## CD8+</span></span>
<span id="cb32-4"><a href="markers-visualization.html#cb32-4" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;NKG7&#39;</span>,     <span class="do">## NK</span></span>
<span id="cb32-5"><a href="markers-visualization.html#cb32-5" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;MS4A1&#39;</span>,    <span class="do">## B cell marker</span></span>
<span id="cb32-6"><a href="markers-visualization.html#cb32-6" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;FCGR3A&#39;</span>,   <span class="do">## Mono</span></span>
<span id="cb32-7"><a href="markers-visualization.html#cb32-7" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;FCER1A&#39;</span>,   <span class="do">## DC</span></span>
<span id="cb32-8"><a href="markers-visualization.html#cb32-8" aria-hidden="true" tabindex="-1"></a>                       <span class="st">&#39;CD14&#39;</span>     <span class="do">## Mono</span></span>
<span id="cb32-9"><a href="markers-visualization.html#cb32-9" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb32-10"><a href="markers-visualization.html#cb32-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-11"><a href="markers-visualization.html#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(pbmc.seurat, </span>
<span id="cb32-12"><a href="markers-visualization.html#cb32-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> canonical_markers,</span>
<span id="cb32-13"><a href="markers-visualization.html#cb32-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="course_files/figure-html/umap_markers_vis-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now, we will annotate the cells with their identified identities in the seurat
object. We will map the cluster names as follows:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="markers-visualization.html#cb33-1" aria-hidden="true" tabindex="-1"></a>mapping <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">seurat_cluster=</span><span class="fu">c</span>(<span class="st">&#39;0&#39;</span>, <span class="st">&#39;1&#39;</span>, <span class="st">&#39;2&#39;</span>, <span class="st">&#39;3&#39;</span>),</span>
<span id="cb33-2"><a href="markers-visualization.html#cb33-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">cell_type=</span><span class="fu">c</span>(<span class="st">&#39;Lymphocyte&#39;</span>, </span>
<span id="cb33-3"><a href="markers-visualization.html#cb33-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">&#39;B cell&#39;</span>, </span>
<span id="cb33-4"><a href="markers-visualization.html#cb33-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">&#39;NK&#39;</span>,</span>
<span id="cb33-5"><a href="markers-visualization.html#cb33-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">&#39;Monocyte&#39;</span>))</span>
<span id="cb33-6"><a href="markers-visualization.html#cb33-6" aria-hidden="true" tabindex="-1"></a>mapping</span></code></pre></div>
<pre><code>##   seurat_cluster  cell_type
## 1              0 Lymphocyte
## 2              1     B cell
## 3              2         NK
## 4              3   Monocyte</code></pre>
<p>To assign the new labels we can use the map function from the plyr R package
as follows:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="markers-visualization.html#cb35-1" aria-hidden="true" tabindex="-1"></a>pbmc.seurat<span class="sc">$</span><span class="st">&#39;cell_type&#39;</span> <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb35-2"><a href="markers-visualization.html#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> pbmc.seurat<span class="sc">$</span>seurat_clusters,</span>
<span id="cb35-3"><a href="markers-visualization.html#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> mapping<span class="sc">$</span>seurat_cluster,</span>
<span id="cb35-4"><a href="markers-visualization.html#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> mapping<span class="sc">$</span>cell_type</span>
<span id="cb35-5"><a href="markers-visualization.html#cb35-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Now, we can plot the clusters with the assigned cell types.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="markers-visualization.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc.seurat, </span>
<span id="cb36-2"><a href="markers-visualization.html#cb36-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">&#39;cell_type&#39;</span>,   <span class="do">## set the column to use as category</span></span>
<span id="cb36-3"><a href="markers-visualization.html#cb36-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>)  <span class="sc">+</span>          <span class="do">## label clusters</span></span>
<span id="cb36-4"><a href="markers-visualization.html#cb36-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">NoLegend</span>()                <span class="do">## remove legends</span></span></code></pre></div>
<p><img src="course_files/figure-html/cell_types-1.png" width="672" style="display: block; margin: auto;" /></p>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-expression-analysis.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
