<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Differential Expression Analysis | IRTG COURSE: INTRODUCTION TO R</title>
  <meta name="description" content="8 Differential Expression Analysis | IRTG COURSE: INTRODUCTION TO R" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Differential Expression Analysis | IRTG COURSE: INTRODUCTION TO R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Differential Expression Analysis | IRTG COURSE: INTRODUCTION TO R" />
  
  
  

<meta name="author" content="Carl Herrmann &amp; Carlos Ramirez" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="cluster-visualization.html"/>
<link rel="next" href="markers-visualization.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> scRNA-Seq Course Syllabus</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#single-cell-sequencing"><i class="fa fa-check"></i><b>1.1</b> Single Cell Sequencing</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#quantifying-gene-expression-in-single-cells"><i class="fa fa-check"></i><b>1.2</b> Quantifying gene expression in single cells</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#quizzes"><i class="fa fa-check"></i><b>1.3</b> Quizzes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html"><i class="fa fa-check"></i><b>2</b> Standard Preprocessing using Seurat</a>
<ul>
<li class="chapter" data-level="2.1" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#creating-seurat-object"><i class="fa fa-check"></i><b>2.1</b> Creating Seurat object</a></li>
<li class="chapter" data-level="2.2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#exploring-the-seurat-object"><i class="fa fa-check"></i><b>2.2</b> Exploring the Seurat object</a></li>
<li class="chapter" data-level="2.3" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#extracting-expression-values"><i class="fa fa-check"></i><b>2.3</b> Extracting expression values</a></li>
<li class="chapter" data-level="2.4" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#quizzes-1"><i class="fa fa-check"></i><b>2.4</b> Quizzes</a></li>
<li class="chapter" data-level="2.5" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#exercises"><i class="fa fa-check"></i><b>2.5</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality control</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#filtering-out-cells"><i class="fa fa-check"></i><b>3.1</b> Filtering out cells</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#quizzes-2"><i class="fa fa-check"></i><b>3.2</b> Quizzes</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#exercises-1"><i class="fa fa-check"></i><b>3.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>4</b> Feature selection</a>
<ul>
<li class="chapter" data-level="4.1" data-path="feature-selection.html"><a href="feature-selection.html#exercises-2"><i class="fa fa-check"></i><b>4.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html"><i class="fa fa-check"></i><b>5</b> Dimensional Reduction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#normalization"><i class="fa fa-check"></i><b>5.1</b> Normalization</a></li>
<li class="chapter" data-level="5.2" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#dimensional-reduction-1"><i class="fa fa-check"></i><b>5.2</b> Dimensional reduction</a></li>
<li class="chapter" data-level="5.3" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#exercises-3"><i class="fa fa-check"></i><b>5.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="cell-clustering.html"><a href="cell-clustering.html"><i class="fa fa-check"></i><b>6</b> Cell clustering</a>
<ul>
<li class="chapter" data-level="6.1" data-path="cell-clustering.html"><a href="cell-clustering.html#exercises-4"><i class="fa fa-check"></i><b>6.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cluster-visualization.html"><a href="cluster-visualization.html"><i class="fa fa-check"></i><b>7</b> Cluster visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cluster-visualization.html"><a href="cluster-visualization.html#umap"><i class="fa fa-check"></i><b>7.1</b> UMAP</a></li>
<li class="chapter" data-level="7.2" data-path="cluster-visualization.html"><a href="cluster-visualization.html#exercise"><i class="fa fa-check"></i><b>7.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential Expression Analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html#exercises-5"><i class="fa fa-check"></i><b>8.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="markers-visualization.html"><a href="markers-visualization.html"><i class="fa fa-check"></i><b>9</b> Markers visualization</a>
<ul>
<li class="chapter" data-level="9.1" data-path="markers-visualization.html"><a href="markers-visualization.html#visualization-of-gene-expression-levels-of-markers-in-clusters"><i class="fa fa-check"></i><b>9.1</b> Visualization of gene expression levels of markers in clusters</a></li>
<li class="chapter" data-level="9.2" data-path="markers-visualization.html"><a href="markers-visualization.html#exercises-6"><i class="fa fa-check"></i><b>9.2</b> Exercises</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">IRTG COURSE: INTRODUCTION TO R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="differential-expression-analysis" class="section level1" number="8">
<h1><span class="header-section-number">8</span> Differential Expression Analysis</h1>
<p>The main advantage of using scRNA-Seq technologies is the possibility of
assessing cell type specificity and heterogeneity, which is not possible while
using bulk assays.</p>
<p>We should expect that some of the identified clusters in the UMAP might correspond
to distinct cell types. The assignation of cell types identities is not always
straightforward, some clusters might still contain some variability, and
additionally, different clusters might correspond to the same cell type at a
different functional, metabolic or cycling point.</p>
<p>Cell type profiling is generally done by assessing the expression of markers.
This task can be done manually by inspecting markers in dimensional reduced data
projected in UMAP or tSNE. It can also be done in a automatic manner scoring
cells using gene signatures, which are lists of marker genes. Scores are generally
based in the median expression of all the markers. Using scores has the advantage
or reducing bias due to the arbitrary selection of markers.</p>
<p>Finding differential expressed markers is important for cluster profiling and
identification. We will use the <code>FindAllMarkers()</code> function, which performs
a statistical test comparing the distribution of gene expression values for
each gene separately comparing one assigned cell type cluster (in this case
the <code>seurat_clusters</code> column) <em>vs</em> the rest of cells.</p>
<p>First, we will set up the column used to define the clusters using the
function <code>Idents()</code>.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="differential-expression-analysis.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pbmc.filtered) <span class="ot">&lt;-</span> pbmc.filtered<span class="sc">$</span>seurat_clusters</span></code></pre></div>
<p>Now we can calculate the DEGs.
There are several parameters for <code>FindAllMarkers()</code>, we will discuss
<code>logfc.threshold</code>, <code>min.pct</code> and <code>min.cells.feature</code> that corresponds to the threshold of gene
expression fold change, the minimum percentage of cells expressing the marker
and the minimum of cells expressing (counts &gt; 0) the feature. These parameters
are used to filter out genes prior calculating DEGs. Lowering the values of these
parameters will increase the sensibility of the method at the expense of
increasing computation time.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="differential-expression-analysis.html#cb35-1" aria-hidden="true" tabindex="-1"></a>pbmc.degs <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(pbmc.filtered, </span>
<span id="cb35-2"><a href="differential-expression-analysis.html#cb35-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">logfc.threshold =</span> <span class="dv">1</span>, </span>
<span id="cb35-3"><a href="differential-expression-analysis.html#cb35-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.pct =</span> <span class="fl">0.05</span>, </span>
<span id="cb35-4"><a href="differential-expression-analysis.html#cb35-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells.feature =</span> <span class="dv">10</span>, </span>
<span id="cb35-5"><a href="differential-expression-analysis.html#cb35-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The output <code>pbmc.degs</code> consist of a data frame contanning the DEGs with
p-vales, p-adjusted values and log fold change values for each gene as
we can see next:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="differential-expression-analysis.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pbmc.degs)</span></code></pre></div>
<pre><code>##                 p_val avg_log2FC pct.1 pct.2    p_val_adj cluster     gene
## HLA-DRA  1.355148e-67  -4.415537 0.300 0.968 1.717379e-63       0  HLA-DRA
## HLA-DRB1 9.575850e-60  -3.310077 0.148 0.884 1.213547e-55       0 HLA-DRB1
## CD74     3.415369e-51  -3.122109 0.781 0.974 4.328297e-47       0     CD74
## HLA-DPB1 1.613117e-46  -3.269898 0.310 0.858 2.044303e-42       0 HLA-DPB1
## IL32     9.357132e-44   3.271671 0.808 0.110 1.185829e-39       0     IL32
## HLA-DRB5 2.825316e-43  -2.749389 0.084 0.684 3.580523e-39       0 HLA-DRB5</code></pre>
<p>We can make a vulcano plot using <code>ggplot</code>:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="differential-expression-analysis.html#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb38-2"><a href="differential-expression-analysis.html#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)         <span class="do">## for handling data frames</span></span>
<span id="cb38-3"><a href="differential-expression-analysis.html#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb38-4"><a href="differential-expression-analysis.html#cb38-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-5"><a href="differential-expression-analysis.html#cb38-5" aria-hidden="true" tabindex="-1"></a>pbmc.degs <span class="sc">%&gt;%</span></span>
<span id="cb38-6"><a href="differential-expression-analysis.html#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(<span class="fu">abs</span>(avg_log2FC))) <span class="sc">%&gt;%</span>       <span class="do">## Arranging genes by FC</span></span>
<span id="cb38-7"><a href="differential-expression-analysis.html#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rank=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(pbmc.degs)) <span class="sc">%&gt;%</span>        <span class="do">## Ranking markers by FC</span></span>
<span id="cb38-8"><a href="differential-expression-analysis.html#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">highlight=</span><span class="fu">ifelse</span>(rank<span class="sc">&lt;</span><span class="dv">20</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>)) <span class="sc">%&gt;%</span> <span class="do">## highlighting top FC markers</span></span>
<span id="cb38-9"><a href="differential-expression-analysis.html#cb38-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_label=</span><span class="fu">ifelse</span>(highlight<span class="sc">==</span><span class="cn">TRUE</span>, gene, <span class="st">&#39;&#39;</span>)) <span class="sc">%&gt;%</span> <span class="do">## Adding labels for top markers</span></span>
<span id="cb38-10"><a href="differential-expression-analysis.html#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>avg_log2FC, <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(p_val_adj),</span>
<span id="cb38-11"><a href="differential-expression-analysis.html#cb38-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">colour=</span>highlight,</span>
<span id="cb38-12"><a href="differential-expression-analysis.html#cb38-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">label=</span>gene_label)) <span class="sc">+</span>         <span class="do">## adding labels for top markers</span></span>
<span id="cb38-13"><a href="differential-expression-analysis.html#cb38-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb38-14"><a href="differential-expression-analysis.html#cb38-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_text_repel</span>() <span class="sc">+</span></span>
<span id="cb38-15"><a href="differential-expression-analysis.html#cb38-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="course_files/figure-html/vulcano_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="exercises-5" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Exercises</h2>
<blockquote>
<p>Compare the DEGs from the above shown results with that calculated using the 10X PBMC 250 cells downsample data clusters in the previous exercises</p>
</blockquote>
<ul>
<li><p>Load the tsv file containing the calculated DEGs from above using the following
url:
<code>https://raw.githubusercontent.com/caramirezal/caramirezal.github.io/master/bookdown-minimal/data/degs_10x_pbmc.tsv</code></p></li>
<li><p>Calculate the DEGs using the downsampled 10X PBMC 250 cells data with the clusters
obtained using your parameters</p></li>
<li><p>Intersect both lists of genes</p></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="cluster-visualization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="markers-visualization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
