<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>IRTG COURSE: INTRODUCTION TO R</title>
  <meta name="description" content="IRTG COURSE: INTRODUCTION TO R" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="IRTG COURSE: INTRODUCTION TO R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="IRTG COURSE: INTRODUCTION TO R" />
  
  
  

<meta name="author" content="Carl Herrmann &amp; Carlos Ramirez" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="standard-preprocessing-using-seurat.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> scRNA-Seq Course Syllabus</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#single-cell-sequencing"><i class="fa fa-check"></i><b>1.1</b> Single Cell Sequencing</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#quantifying-gene-expression-in-single-cells"><i class="fa fa-check"></i><b>1.2</b> Quantifying gene expression in single cells</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html"><i class="fa fa-check"></i><b>2</b> Standard Preprocessing using Seurat</a>
<ul>
<li class="chapter" data-level="2.1" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#creating-seurat-object"><i class="fa fa-check"></i><b>2.1</b> Creating Seurat object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality control</a></li>
<li class="chapter" data-level="4" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>4</b> Feature selection</a></li>
<li class="chapter" data-level="5" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html"><i class="fa fa-check"></i><b>5</b> Dimensionional Reduction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html#normalization"><i class="fa fa-check"></i><b>5.1</b> Normalization</a></li>
<li class="chapter" data-level="5.2" data-path="dimensionional-reduction.html"><a href="dimensionional-reduction.html#dimensional-reduction"><i class="fa fa-check"></i><b>5.2</b> Dimensional reduction</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="cell-clustering.html"><a href="cell-clustering.html"><i class="fa fa-check"></i><b>6</b> Cell clustering</a></li>
<li class="chapter" data-level="7" data-path="cluster-visualization.html"><a href="cluster-visualization.html"><i class="fa fa-check"></i><b>7</b> Cluster visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cluster-visualization.html"><a href="cluster-visualization.html#umap"><i class="fa fa-check"></i><b>7.1</b> UMAP</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="9" data-path="markers-visualization.html"><a href="markers-visualization.html"><i class="fa fa-check"></i><b>9</b> Markers visualization</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">IRTG COURSE: INTRODUCTION TO R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">IRTG COURSE: INTRODUCTION TO R</h1>
<p class="author"><em>Carl Herrmann &amp; Carlos Ramirez</em></p>
<p class="date" style="margin-top: 1.5em;"><em>08 June, 2021</em></p>
</div>
<div id="intro" class="section level1" number="1">
<h1><span class="header-section-number">1</span> scRNA-Seq Course Syllabus</h1>
<p>We will use available data from 10X genomics of Peripheral Blood Mononuclear cells (PBMC). In order
to speed up the process only a subset of 500 cells are going to be processed here. Here, we will
perform the following tasks:</p>
<ul>
<li>Definition of a Seurat object from a matrix of counts</li>
<li>Quality Control (QC) of the cell samples</li>
<li>Filtering</li>
<li>Dimensional reduction</li>
<li>Clustering</li>
<li>Differential expression</li>
<li>Visualization of markers</li>
</ul>
<div id="single-cell-sequencing" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Single Cell Sequencing</h2>
<p>Single cell sequencing (scRNA-Seq) technologies arise from bulk counterparts with
the aim of refining gene expression profiles. Previous bulk Sequencing offered
averaged quantifications of gene expression in samples. In some contexts, for
example, when studying cell type specificity or the heterogeneity in tumours is
important to dissect patterns in cell subpopulations.</p>
<p>The following image illustrates the development of scRNA-Seq technologies
and the number of cells that can be analyzed with each technology. In this
course we will not cover this topic, but instead we will take a hands on
approach and jump directly to the analysis of scRNA-Seq data. However, the
analysis that we show next are applied downstream to libraries construction
and sequencing and are therefore platform agnostic.</p>
<div class="figure">
<img src="figures/moores-law.png" alt="" />
<p class="caption"><strong>Single Cell Sequencing Platforms</strong>: Date of development <em>vs</em> number of cells analyzed by each technology.</p>
</div>
<p>Image is taken from <a href="https://arxiv.org/abs/1704.01379">Svensson V et al, 2017</a>.</p>
</div>
<div id="quantifying-gene-expression-in-single-cells" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Quantifying gene expression in single cells</h2>
<p>Genes are expressed in cells by the synthesis of mRNA molecules. Quantification of gene expression
in single cells consists of counting and mapping sequenced reads to each gene in each cell. The
outcome of this process is a count matrix, which is a discrete matrix of integers corresponding
to the number of RNA molecules associated to each gene and cell. The count matrix usually has
genes as rows and cells in columns.</p>
<p>In the next lines we will load the matrix of UMI counts into a variable called <code>pbmc.mtx</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Dependencies</span></span>
<span id="cb1-2"><a href="index.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rmarkdown)</span>
<span id="cb1-3"><a href="index.html#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="index.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">333</span>)</span>
<span id="cb1-5"><a href="index.html#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="index.html#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Loading example data</span></span>
<span id="cb1-7"><a href="index.html#cb1-7" aria-hidden="true" tabindex="-1"></a>pbmc_url  <span class="ot">&lt;-</span> <span class="st">&#39;https://github.com/caramirezal/caramirezal.github.io/raw/master/bookdown-minimal/data/pbmc_10X_500_cells.mtx.rds&#39;</span></span>
<span id="cb1-8"><a href="index.html#cb1-8" aria-hidden="true" tabindex="-1"></a>pbmc.mtx <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(pbmc_url))</span></code></pre></div>
<p>As can be seen from the output of the function <code>dim(pbmc.mtx)</code> the matrix contains rows (genes) and 500 columns (cells).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pbmc.mtx)</span></code></pre></div>
<pre><code>## [1] 13714   500</code></pre>
<p>To have a glimpse of the appearance of this matrix we print the first 5 rows and first 5 columns. It can
be seen that the matrix object is a particular array called sparse Matrix. These types of matrices are
specially efficient to store numerical arrays which are very sparse (full of zeroes) as is the case
for single cell count matrices data.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="index.html#cb4-1" aria-hidden="true" tabindex="-1"></a>pbmc.mtx[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
##               AAAGAGACGGACTT-1 AAAGATCTGGGCAA-1 AAAGCAGATATCGG-1
## AL627309.1                   .                .                .
## AP006222.2                   .                .                .
## RP11-206L10.2                .                .                .
## RP11-206L10.9                .                .                .
## LINC00115                    .                .                .
##               AAAGTTTGATCACG-1 AAATCAACCCTATT-1
## AL627309.1                   .                .
## AP006222.2                   .                .
## RP11-206L10.2                .                .
## RP11-206L10.9                .                .
## LINC00115                    .                .</code></pre>
<p>We can start from different types of arrays apart from sparse matrices as data frames or hd5
objects just to give some examples. The format in which the matrix is provided might also vary from
excel files, tsv, csv, rds, etc. Usually this formats can be converted into each other, however, the
methods for such transformation are so different and specific that we cannot covered here.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="standard-preprocessing-using-seurat.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
