<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Standard Preprocessing using Seurat | IRTG COURSE: INTRODUCTION TO R</title>
  <meta name="description" content="2 Standard Preprocessing using Seurat | IRTG COURSE: INTRODUCTION TO R" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Standard Preprocessing using Seurat | IRTG COURSE: INTRODUCTION TO R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Standard Preprocessing using Seurat | IRTG COURSE: INTRODUCTION TO R" />
  
  
  

<meta name="author" content="Carl Herrmann &amp; Carlos Ramirez" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="quality-control.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> scRNA-Seq Course Syllabus</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#single-cell-sequencing"><i class="fa fa-check"></i><b>1.1</b> Single Cell Sequencing</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#quantifying-gene-expression-in-single-cells"><i class="fa fa-check"></i><b>1.2</b> Quantifying gene expression in single cells</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#quizzes"><i class="fa fa-check"></i><b>1.3</b> Quizzes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html"><i class="fa fa-check"></i><b>2</b> Standard Preprocessing using Seurat</a>
<ul>
<li class="chapter" data-level="2.1" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#creating-seurat-object"><i class="fa fa-check"></i><b>2.1</b> Creating Seurat object</a></li>
<li class="chapter" data-level="2.2" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#exploring-the-seurat-object"><i class="fa fa-check"></i><b>2.2</b> Exploring the Seurat object</a></li>
<li class="chapter" data-level="2.3" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#extracting-expression-values"><i class="fa fa-check"></i><b>2.3</b> Extracting expression values</a></li>
<li class="chapter" data-level="2.4" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#quizzes-1"><i class="fa fa-check"></i><b>2.4</b> Quizzes</a></li>
<li class="chapter" data-level="2.5" data-path="standard-preprocessing-using-seurat.html"><a href="standard-preprocessing-using-seurat.html#exercises"><i class="fa fa-check"></i><b>2.5</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality control</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#filtering-out-cells"><i class="fa fa-check"></i><b>3.1</b> Filtering out cells</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#quizzes-2"><i class="fa fa-check"></i><b>3.2</b> Quizzes</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#exercises-1"><i class="fa fa-check"></i><b>3.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>4</b> Feature selection</a>
<ul>
<li class="chapter" data-level="4.1" data-path="feature-selection.html"><a href="feature-selection.html#exercises-2"><i class="fa fa-check"></i><b>4.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html"><i class="fa fa-check"></i><b>5</b> Dimensional Reduction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#normalization"><i class="fa fa-check"></i><b>5.1</b> Normalization</a></li>
<li class="chapter" data-level="5.2" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#dimensional-reduction-1"><i class="fa fa-check"></i><b>5.2</b> Dimensional reduction</a></li>
<li class="chapter" data-level="5.3" data-path="dimensional-reduction.html"><a href="dimensional-reduction.html#exercises-3"><i class="fa fa-check"></i><b>5.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="cell-clustering.html"><a href="cell-clustering.html"><i class="fa fa-check"></i><b>6</b> Cell clustering</a>
<ul>
<li class="chapter" data-level="6.1" data-path="cell-clustering.html"><a href="cell-clustering.html#exercises-4"><i class="fa fa-check"></i><b>6.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cluster-visualization.html"><a href="cluster-visualization.html"><i class="fa fa-check"></i><b>7</b> Cluster visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cluster-visualization.html"><a href="cluster-visualization.html#umap"><i class="fa fa-check"></i><b>7.1</b> UMAP</a></li>
<li class="chapter" data-level="7.2" data-path="cluster-visualization.html"><a href="cluster-visualization.html#exercise"><i class="fa fa-check"></i><b>7.2</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential Expression Analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="differential-expression-analysis.html"><a href="differential-expression-analysis.html#exercises-5"><i class="fa fa-check"></i><b>8.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="markers-visualization.html"><a href="markers-visualization.html"><i class="fa fa-check"></i><b>9</b> Markers visualization</a>
<ul>
<li class="chapter" data-level="9.1" data-path="markers-visualization.html"><a href="markers-visualization.html#visualization-of-gene-expression-levels-of-markers-in-clusters"><i class="fa fa-check"></i><b>9.1</b> Visualization of gene expression levels of markers in clusters</a></li>
<li class="chapter" data-level="9.2" data-path="markers-visualization.html"><a href="markers-visualization.html#exercises-6"><i class="fa fa-check"></i><b>9.2</b> Exercises</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">IRTG COURSE: INTRODUCTION TO R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="standard-preprocessing-using-seurat" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Standard Preprocessing using Seurat</h1>
<p>Some standard steps are usually carried out in scRNA-Seq prior to further analysis as QC, dimensional
reduction and marker visualization. Here, we will use the Seurat R package to perform these steps which
is increasingly becoming the most popular tool, however, there are some other options as SingleCellExperiment
in R and scanpy available for python. First, we need to define a Seurat object.</p>
<div id="creating-seurat-object" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Creating Seurat object</h2>
<p>We create a Seurat object using the <code>CreateSeuratObject</code> function as follows. Use the <code>?function</code> helper
in R to get information about the parameters that are need to be provided to the function.</p>
<ul>
<li>counts - a count matrix. It can be a matrix, sparse matrix or dataframe.</li>
<li>project - A single character string. Correspond to a arbitrary name to label the object.</li>
<li>assay - A single character string. An arbitrary name that usually is assigned to label the
type of sequencing information, for example, RNA, spliced RNA, ATAC, etc…</li>
<li>min.cells - An integer. Indicates a threshold of the number of cells for which a feature was
recorded. Cells below that threshold will be filtered out.</li>
<li>min.features - An integer. Similar to min cells but for the number of features of a cell.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="standard-preprocessing-using-seurat.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb6-2"><a href="standard-preprocessing-using-seurat.html#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="standard-preprocessing-using-seurat.html#cb6-3" aria-hidden="true" tabindex="-1"></a>pbmc.seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb6-4"><a href="standard-preprocessing-using-seurat.html#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> pbmc.mtx, </span>
<span id="cb6-5"><a href="standard-preprocessing-using-seurat.html#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">&#39;PBMC&#39;</span>, </span>
<span id="cb6-6"><a href="standard-preprocessing-using-seurat.html#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">&#39;RNA&#39;</span>, </span>
<span id="cb6-7"><a href="standard-preprocessing-using-seurat.html#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.cells =</span> <span class="dv">1</span>, </span>
<span id="cb6-8"><a href="standard-preprocessing-using-seurat.html#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.features =</span> <span class="dv">1</span></span>
<span id="cb6-9"><a href="standard-preprocessing-using-seurat.html#cb6-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The variable <code>pbmc.seurat</code> now contains the Seurat object that we can feed into the package.
If we print the variable we get information about the number of genes and cells.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="standard-preprocessing-using-seurat.html#cb7-1" aria-hidden="true" tabindex="-1"></a>pbmc.seurat</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 12673 features across 500 samples within 1 assay 
## Active assay: RNA (12673 features, 0 variable features)</code></pre>
</div>
<div id="exploring-the-seurat-object" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Exploring the Seurat object</h2>
<p>Seurat objects can be seen as a container of different features. At this step it contains
our gene expression matrix, but in addition it can store metadata, processed data,
information from different assays, for example, scATACSeq, scCITESeq or unspliced transcripts.</p>
<p>We can explore the seurat object using the <code>$</code> to explore its <em>metadata</em> in combination with the tab
key. For example, during the creation of the seurat object the number of counts quality metric
is calculated and added to the metadata. We can explore this metric by accessing the metadata
as follows.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="standard-preprocessing-using-seurat.html#cb9-1" aria-hidden="true" tabindex="-1"></a>pbmc.seurat<span class="sc">$</span>nCount_RNA <span class="sc">%&gt;%</span> head</span></code></pre></div>
<pre><code>## AAAGAGACGGACTT-1 AAAGATCTGGGCAA-1 AAAGCAGATATCGG-1 AAAGTTTGATCACG-1 
##             1151             1347             4584             1268 
## AAATCAACCCTATT-1 AAATGTTGTGGCAT-1 
##             5676             2761</code></pre>
<p>We can do the same with the <code>@</code> operator to explore the different <em>slots</em>. For example,
we can extract the original count matrix that we used to create the seurat object as follows:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="standard-preprocessing-using-seurat.html#cb11-1" aria-hidden="true" tabindex="-1"></a>pbmc.seurat<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
##            AAAGAGACGGACTT-1 AAAGATCTGGGCAA-1 AAAGCAGATATCGG-1 AAAGTTTGATCACG-1
## AL627309.1                .                .                .                .
## AP006222.2                .                .                .                .
## LINC00115                 .                .                .                .
## NOC2L                     .                .                .                .
## PLEKHN1                   .                .                .                .
##            AAATCAACCCTATT-1
## AL627309.1                .
## AP006222.2                .
## LINC00115                 .
## NOC2L                     .
## PLEKHN1                   .</code></pre>
</div>
<div id="extracting-expression-values" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Extracting expression values</h2>
<p>Next, we visualize gene counts to see its behavior. We take a look at the expression of the
house keeping gene ACTIN Beta and plot an histogram of count values. We will use the
the function <code>FetchData</code> which is used to extract values from selected features in the Seurat
object and then plot it using an histogram.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="standard-preprocessing-using-seurat.html#cb13-1" aria-hidden="true" tabindex="-1"></a>actin <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(pbmc.seurat, <span class="at">vars =</span> <span class="st">&#39;ACTB&#39;</span>)</span>
<span id="cb13-2"><a href="standard-preprocessing-using-seurat.html#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(actin<span class="sc">$</span>ACTB)</span></code></pre></div>
<p><img src="course_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="quizzes-1" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Quizzes</h2>
<br>
<details>
<summary>
<em>Found and display the metadata in the seurat object</em>
<br>
TIP: You can have
a look at the <a href="https://github.com/satijalab/seurat/wiki/Seurat#object-information">documentation</a>
of the seurat objects from the GitHub Wiki.
</summary>
<br>
<b>Answer:</b>
<br>
<code>pbmc.seurat@meta.data %&gt;% head</code>
</details>
</div>
<div id="exercises" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Exercises</h2>
<blockquote>
<p>Create a Seurat object</p>
</blockquote>
<p>The file in the follwing URL:</p>
<p><code>https://raw.githubusercontent.com/caramirezal/caramirezal.github.io/master/bookdown-minimal/data/pbmc_10X_250_cells.tsv</code></p>
<p>contains 250 cells downsampled from the 10x PBMC data and stored in tsv format</p>
<ul>
<li>Load the count matrix in tsv format</li>
<li>Create a Seurat object using the count matrix</li>
<li>How many features and cells are present in the count matrix?</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quality-control.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
